### PR TITLE
boot: Fix default/oneshot selection for EFISTUB entries

### PR DESCRIPTION
**Pull request by SpencerMichaels**
# Summary

This PR fixes a behavior in `systemd-boot` in which setting a Linux EFISTUB boot entry as the _default_ or _oneshot_ entry can result in booting an entry other than the one selected, especially when multiple entries exist for the same distro (e.g. when multiple EFISTUBs for different kernel versions or configurations are present). This issue is caused by default/oneshot patterns being derived from, and matched against, non-unique ID values. It is fixed by changing the way EFISTUB IDs are calculated, making entry IDs globally unique.

# Problem

`systemd-boot`'s current behavior for determining entry IDs is as follows:

- **Loader configs**, i.e. `/loader/entries/*.conf` files: Each entry's `id` value is simply the entry's filename, minus the `.conf` suffix. See `config_entry_add_from_file()` at [boot.c:1433](src/boot/efi/boot.c#L1433).
- **Linux EFISTUB binaries**, i.e. `/EFI/Linux/*.efi` files: Each entry's `id` value is derived from a combination of the `ID`, `VERSION_ID`, and `BUILD_ID` values in the `/etc/os-release` file embedded in the binary. Specifically, the final value is `{id}-{version}` if a `VERSION_ID` value is defined, and `{id}-{build}` otherwise. See `config_entry_add_linux()` at [boot.c:1941](src/boot/efi/boot.c#L1941).

A major flaw in the latter method is that values pulled from `/etc/os-release` are likely to be identical across completely different kernel versions and configurations, so long as the distro version is the same. (The latest release of Ubuntu, for instance, sets `ID=ubuntu` and `VERSION_ID=19.04`; Arch Linux uses static values of `ID=arch` and `BUILD_ID=rolling`.) This results in multiple EFISTUB entries for a given distro release having exactly the same ID, even if their contents are different.

If a default or oneshot entry is set, `systemd-boot` will match the saved pattern — which will be the ID of the selected entry, or a prefix of it — against the IDs of each entry in turn, and pick the first one that matches. (See `config_default_entry_select()` at [boot.c:1589]( src/boot/efi/boot.c#L1589).) As such, in the likely case that multiple entries have the same ID, `systemd-boot` will only ever boot from the first of them, even if the second, third, etc. was selected.

# Solution

- Change `config_entry_add_linux()` so that Linux EFISTUB entries' `id`s are simply the binaries' filenames, as is already the case with loader configs.  This guarantees that EFISTUB binaries' IDs are mutually unique.
  - To further increase similarity with loader configs and provide global uniqueness, ignore filenames starting with `auto-` (see [boot.c:1512](src/boot/efi/boot.c#L1512)).
- Change `config_entry_add_from_file()` so that it does not remove the `.conf` ending from the filename that it uses for `entry->id`.
  - This is necessary because otherwise an EFISTUB binary and a loader config might end up with the same ID, as the loader config IDs don't include the config filename's `.conf` extension (see [boot.c:1435](src/boot/efi/boot.c#L1435)).  Consider, for instance, an EFISTUB `/EFI/Linux/linux.efi` and a loader config `/loader/entries/linux.efi.conf`; both would have the ID `linux.efi`.
  
The above is sufficient to guarantee that all entry IDs are unique, including those of auto-detected entries for non-Linux OSes, e.g. `auto-osx`. This is because the following invariants are upheld:

- EFISTUB entry IDs are mutually unique (no two files in `/EFI/Linux` can have the same name), always end in `.efi`, and cannot start with `auto-`.
- Loader entry IDs are mutually unique (no two files in `/loader/entries/linux.conf` can have the same name), always end in `.conf`, and cannot start with `auto-`.
- Autodetected non-Linux EFI entries all start with `auto-` and do not have an extension (see `config_entry_add_osx()` at [boot.c:1872](src/boot/efi/boot.c#L1872) and `efi_main()` at [boot.c:2410](src/boot/efi/boot.c#L2410)).

In addition to the above, I made two minor cosmetic changes to how `config_entry_add_linux()` extracts OS information from the embedded `/etc/os-release` in EFISTUB binaries. This increases the detail of information available to the user for EFISTUB entries.
- `entry->version`, which was previously always null, is now set to the value of `VERSION`, or `VERSION_ID`, or `BUILD_ID` (preferred in that order).
- `entry->title`, which was previously set to the value of `PRETTY_NAME`, can now fall back to `NAME` or `ID` if `PRETTY_NAME` is not present.  In that case, `NAME` is preferred over `ID`.

# Testing

There are currently no automated tests for `systemd-boot`. Since my changes are relatively minor, however, I was able to test the affected behavior manually and verify that the patch works as intended, and that there do not appear to be any regressions.

# Preview

Below is the output of the `P` command with two entries present: `/loader/entries/arch.conf` and `/EFI/Linux/arch.efi`. For testing purposes, I changed the embedded `/etc/os-release` of the latter from the default one provided by Arch Linux to (a) make the `PRETTY_NAME` and `NAME` values distinct, as they were previously identical, and (b) add a `VERSION` value, as Arch does not provide one.

ss

### CODE REVIEW

**Comment by SpencerMichaels** [comment]
Hm, not sure why `autopkgtest` has failed on `bionic-ppc64el` above — perhaps just a sporadic error? The error in question (line 120531) appears to be a page fault after the system has booted, which as far as I know is unrelated to anything my changes to `systemd-boot` could have induced.

```
OK  [0m] Reached target [Login Prompts]
systemd-testsuite login: [  436.497878] Unable to handle kernel paging request for data at address 0x7244628f59602071
```

Please let me know if I should re-push to run the tests again, or otherwise need to change anything.

**Comment by SpencerMichaels** [comment]
OK, all tests are passed now. Looks like it was indeed a sporadic error; just had to make CI re-run the tests (for future reference, the following worked without substantially changing commit history: `git commit --amend --no-edit && git push --force`).

File: src/boot/efi/boot.c [hunk 1]
```diff
@@ -1431,10 +1430,6 @@ static VOID config_entry_add_from_file(
 
         entry->device = device;
         entry->id = StrDuplicate(file);
-        len = StrLen(entry->id);
-        /* remove ".conf" */
-        if (len > 5)
-                entry->id[len - 5] = '\0';
```
Comment (keszybz) [inline src/boot/efi/boot.c:1437]: There is matching code in `src/shared/bootspec.c:boot_entry_load`. This same change would have to be done there.

**Comment by poettering** [comment]
Hmm, so EFI binary contents are potentially signed, file names aren't. type 2 boot loader entries are particularly interesting because they are signed. I think it's hence preferable using data from inside the images for identification than from the name hence, since it's protected that way. (of course, we currently don't validate the image when parsing, only when executing, but that sounds like something we might add eventually in some form, and hence maybe keep the semantics already compatible for this to be locked down later)

**Comment by poettering** [comment]
(i.e. maybe one way out is to combine both: use the ids from inside, but suffix them with the file name)

**Comment by SpencerMichaels** [comment]
@keszybz: Thanks! I adjusted the ID computation in `bootspec.c`, and copied the `summary` section above into the long-form commit message of each commit. Is that the way you wanted me to do it? I looked at git log messages from previous commits and they generally did not contain anything as long as my entire PR description, so I was not sure if you meant just the summary.

@poettering: If you are concerned about ensuring that the default/oneshot image to be booted is bit-wise the same image as what was previously selected, it seems to me that the only sure option is to save some kind of hash of the EFISTUB in question. Consider two images that differ only in what binaries are present in their initrd; this would be the case for `fallback` entries, for instance. Detecting that kind of difference without just comparing bits using hashes would be complicated.

Perhaps the bootloader could do the following — there may be a simpler method and I'm not super familiar with bootloaders, so please let me know if there's a better solution I'm not thinking of.

1. Obtain the default/oneshot entry ID from EFI vars, formatted as `<id>[:<hash>]` where the hash component is optional.
  - The `id` component is derived from the filename as this PR currently does.
  - The `:<hash>` ID suffix could be the result of a common hash algorithm on the EFI binary, or maybe [extract](https://github.com/msekletar/sbsigntool/blob/master/src/sbverify.c#L160) the EFISTUB's preexisting signature (GPG signatures are probably too long to store in an EFI variable, so a "digest" may be necessary).
2. Find the entry with a matching ID. If it exists, and is a signed EFISTUB, compute a hash for comparison.
3. If the hash matches that in the saved default/oneshot, boot the entry. Otherwise, either throw up an error, or simply warn the user and refuse to automatically boot that entry, dropping to the regular entry selection menu.

One problem this might introduce is that it would be more difficult to set the default/oneshot entry from _outside_ the bootloader via `bootctl set-oneshot` and `set-default`, which currently just takes the raw ID value. If we just use a standard hash algorithm and document it, that shouldn't be a problem, but if we did something more esoteric based on the EFISTUB signature it might not be easy for end-users to replicate. Although, in the latter case, we could have `bootctl` take a path to a signed EFISTUB and do the hash computation itself.

**Comment by keszybz** [comment]
> Thanks! I adjusted the ID computation in bootspec.c, and copied the summary section above into the long-form commit message of each commit. Is that the way you wanted me to do it? 

Not really. Commits should be self-contained changes and the commit message should describe the what and the why of the commit. Copying the same boilerplate text that talks about the PR is not useful. If the "what" part is simple, then the subject is usually enough to describe that part. But it is still important to describe the "why" part, and that usually happens in the part below the subject. The titles of your commits are OK, but the explanations that should follow are missing. Those parts are already present in the PR description (mostly under Solutions chapter), but they should be split into the individual commits.

Also, the last commit removes a comment. This change should be merged into the commit that makes that comment obsolete.

I find the PR harder to review right now than it should be, because I don't know what your *intent* is in each individual commit.

>  I think it's hence preferable using data from inside the images for identification than from the name hence, since it's protected that way.

Yes, we cannot use the filenames for anything requiring "trust". But I don't think this matters right now. We don't do any kind of verification when parsing entries or selecting the entry to boot, and the way "id" is calculated is largely an internal implementation detail. If the code is changed to do verification, then we can change that then. So I think this PR is an improvement, because it clearly fixes a case where entries are confused by mistake. Making the system resilient to intentional interference is much harder.

>  If you are concerned about ensuring that the default/oneshot image to be booted is bit-wise the same image as what was previously selected, it seems to me that the only sure option is to save some kind of hash of the EFISTUB in question.

That is one case to consider.

Another case would be somebody accessing the EFI partition and simply renaming the .efi files, for example to trick the user into booting an older vulnerable version.

**Comment by SpencerMichaels** [comment]
@k

> > Thanks! I adjusted the ID computation in bootspec.c, and copied the summary section above into the long-form commit message of each commit. Is that the way you wanted me to do it?
> 
> Not really. Commits should be self-contained changes and the commit message should describe the what and the why of the commit. Copying the same boilerplate text that talks about the PR is not useful. If the "what" part is simple, then the subject is usually enough to describe that part. But it is still important to describe the "why" part, and that usually happens in the part below the subject. The titles of your commits are OK, but the explanations that should follow are missing. Those parts are already present in the PR description (mostly under Solutions chapter), but they should be split into the individual commits.

Got it. I had misunderstood what you asked me to do in your first comment; sorry about that. Each commit's long-form comment now contains the relevant subsection of the description. I also rearranged the commits to ensure the descriptions built on each other so they make sense when read in order.

> Also, the last commit removes a comment. This change should be merged into the commit that makes that comment obsolete.

Done. This change is merged into the commit that makes IDs unique (the third in the series); I mentioned the edit in the commit description as well.

>> I think it's hence preferable using data from inside the images for identification than from the name hence, since it's protected that way.
> Yes, we cannot use the filenames for anything requiring "trust". But I don't think this matters right now. We don't do any kind of verification when parsing entries or selecting the entry to boot, and the way "id" is calculated is largely an internal implementation detail. If the code is changed to do verification, then we can change that then. So I think this PR is an improvement, because it clearly fixes a case where entries are confused by mistake. Making the system resilient to intentional interference is much harder.

Sounds reasonable. I would be happy to take a stab at a separate PR to implement this in the future.

> > If you are concerned about ensuring that the default/oneshot image to be booted is bit-wise the same image as what was previously selected, it seems to me that the only sure option is to save some kind of hash of the EFISTUB in question.
> 
> That is one case to consider.
> 
> Another case would be somebody accessing the EFI partition and simply renaming the .efi files, for example to trick the user into booting an older vulnerable version.

That's exactly the case that I was thinking of, actually. If we assume the user is using Secure Boot, then it's already the case that systemd-boot can only boot a fixed set of known-good entries, but an attacker might want a user to mistakenly boot a previous entry with a known vulnerability. The reason I thought that it's probably necessary to check the image bit-for-bit is because there is technically a bunch of software embedded in the initrd that an attacker _might_ be interested in downgrading. More likely, however, the boot target would be either (a) an entry with different command line flags, or (b) an older, known-vulnerable kernel. In the latter case there is actually a pretty straightforward way to check the version: kernel images have an embedded ASCII version string (search `kernel_version` [here](https://www.kernel.org/doc/Documentation/x86/boot.txt)) that I think could just be extracted straight from the `.linux` section an EFISTUB image. That string, concatenated with the command line and then hashed, could provide a reasonable assurance against (what I think are) the two most probable forms of a downgrade attack.

**Comment by keszybz** [comment]
>  In the latter case there is actually a pretty straightforward way to check the version: kernel images have an embedded ASCII version string (search `kernel_version` in [this document](https://www.kernel.org/doc/Documentation/x86/boot.txt) for details) that I think could just be extracted straight from the `.linux` section of an EFISTUB image. That string, hashed in combination with the command line string, could provide a reasonable assurance against (what I think are) the two most probable forms of a downgrade attack, although of course not as much as a full hash of the EFISTUB binary.

But how would we verify the file that contains the hash? It is not signed. If the attacker has access to the EFI partition, they can just hash one of the other entries and replace the file that specifies the default.

**Comment by SpencerMichaels** [comment]
I was thinking that the hash would be stored as part of the
`LoaderEntryDefault` or `LoaderEntryOneshot` EFI variable (e.g. in the
format of `<id>:<hash>`) which is stored in NVRAM, not the filesystem. That
would be harder to edit than the EFI partition filesystem, although
certainly not impossible since we're already positing an attacker with
physical access. It would help if that value could be put behind some kind
of authentication like the Secure Boot keys are, but from a preliminary
search it appears that may not be possible. I'd have to look into UEFI
security further before I could say anything definitive - either way,
though, I think any changes in that area would be orthogonal to this fix
and should be a separate PR.

**Comment by keszybz** [comment]
When I run `bootctl list` with this PR, I get a bunch of additional entries like this:
```
        title: 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.3.7-301.fc31.x86_64
           id: 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.3.7-301.fc31.x86_64
       source: /sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f

        title: 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.3.8-300.fc31.x86_64
           id: 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.3.8-300.fc31.x86_64
       source: /sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f

        title: 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.3.9-300.fc31.x86_64
           id: 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.3.9-300.fc31.x86_64
       source: /sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f
```
It seems to happen because I have the old version of sd-boot installed, so /sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f contains entries with ids without the newly-added ".conf" suffix. This is rather ugly. We don't require people to upgrade their sd-boot installation immediately, and they would see those confusing entries until they do. So maybe `boot_entries_augment_from_loader` should be enhanced to ignore entries which match except for the suffix?

**Comment by SpencerMichaels** [comment]
Hmm, suffix-agnostic matching would work to deduplicate loader config entry names, but not those of EFISTUBs, whose old- and new-style IDs are not guaranteed to be related at all. I can think of two ways to resolve this, both of which have pros and cons:

1. Adopt your solution for removing duplicate loader config entries. In addition, modify the `BootEntry` struct in `bootspec.h` to include a new field, `char *id_old`, which `boot_entry_load_unified()` then populates with the old-style ID calculated from `/etc/os-release`. Then, in `boot_entries_augment_from_loader()`, ignore IDs cached in the `LoaderEntries` EFI var which match the `id_old` field of an existing entry. This achieves perfect deduplication — an entry will be removed if and only if either its old-style or new-style ID is present among the cached entry IDs.
2. Based on the invariant mentioned earlier that all non-auto new-style entries end in either `.conf` or `.efi`, just throw out cached entries from the `LoaderEntries` EFI var that do not fit this invariant. There is an edge case in which this can throw out a non-duplicate ID: if there is an old-style ID cached whose corresponding loader config or EFISTUB has been deleted since last boot, it will not be listed (whereas in previous versions of bootctl it would have been listed).

Method 2 is cleaner in terms of implementation, but may produce behavior that doesn't make sense to people who are in the situation you mentioned. For something as critical as systemd, I suppose that strict backwards compatibility is necessary, so I am pushing a commit implementing Method 1 first. If you'd prefer method 2 and don't think the issue I mentioned is a big deal, let me know and I'll change to that.

**Comment by keszybz** [comment]
Looks good, thanks.

**Pull request merged**